(DOCSP-24575): Update Realm Object Schema page (#2202)
mongodb/docs-realm: source/sdk/flutter/realm-database/update-realm-object-schema.txt

.. _flutter-update-realm-object-schema:
.. _flutter-update-schema:

==========================================

> Update a Realm Object Schema - Flutter SDK

Reviewer: This page is really good but has a lot of links, would it be worth removing some of the duplicate links?

==========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

You can change the schema of a Realm object after you first create it.

> Depending on the changes you make to the schema, the changes can be either

Reviewer: ```suggestion
Depending on the type of changes you make to the schema, the changes can be either
```

To add a property to a schema:

#. Add the new property to the object's ``RealmModel`` class.
#. :ref:`Regenerate the RealmObject <flutter-generate-realm-object>`.
#. Set a schema version to the realm's
   :flutter-sdk:`Configuration object <realm/Configuration-class.html>`.

.. example::

   A realm using schema version ``1`` has a ``Person`` object type with a
   ``firstName``, and ``lastName`` property. The developer decides to add an
   ``age`` property to the ``_Person`` RealmModel class.

   To change the realm to conform to the updated ``Person`` schema, the
   developer sets the realm's :ref:`schema version <flutter-schema-version>`
   to ``2``.

   .. literalinclude:: /examples/generated/flutter/define_realm_model_test.snippet.modify-schema-model.dart
      :language: dart
      :emphasize-lines: 5

   .. literalinclude:: /examples/generated/flutter/define_realm_model_test.snippet.schema-version.dart
      :language: dart

.. _flutter-delete-a-property:

Delete a Property
~~~~~~~~~~~~~~~~~

To delete a property from a schema:

#. Remove the property from the object's ``RealmModel`` class.
#. :ref:`Regenerate the RealmObject <flutter-generate-realm-object>`.

> #. Set a ``schema`` of the realm's ``Configuration`` object.

Author: good spot, even i'm not sure what i meant there. will refactor.

   schema.

   To migrate the realm to conform to the updated ``Dog`` schema, the
   developer sets the realm's :ref:`schema version <flutter-schema-version>` to
   ``2``.

   .. literalinclude:: /examples/generated/flutter/define_realm_model_test.snippet.schema-version.dart
      :language: dart

.. _flutter-manually-migrate-schema:

Manually Migrate Schema
-----------------------

For more complex schema updates, Realm Database requires you to manually migrate
old instances of a given object to the new schema.

When you open the realm with the updated schema, you must do the following:

- Increment the schema version
- Define the migration logic in a :flutter-sdk:`migrationCallback <realm/MigrationCallback.html>`
  property of a realm's :flutter-sdk:`Configuration <realm/Configuration-class.html>`.

The migration callback has the following parameters:

- ``migration``: A :flutter-sdk:`Migration <realm/Migration-class.html>` instance with access to the
  current realm, the realm that you're migrating to, and methods
  to help with other migration tasks.
- ``oldSchemaVersion``: The number of the previous schema version of the realm on
  the device.

> The following sections explain how to perform various types of migration tasks.

Author: i like operation more as well. more precise

.. _flutter-delete-object-type:

Delete an Object Type
~~~~~~~~~~~~~~~~~~~~~

> To delete all objects of a type from your realm, pass a string representation

Author: that is covered a few lines below. i agree that this sentence isn't the most clear. but it's a hard one to represent w/o code. however, we're having a sentence-long overview of all the other migration types on this page, so i thought should here too. 

would advocate keeping this as it is.

To delete all objects of a type from your realm, pass a string representation

> of the an object schema's name to :flutter-sdk:`Migration.deleteType() <realm/Migration/deleteType.html>`.

Reviewer 2: ```suggestion
of the object schema's name to :flutter-sdk:`Migration.deleteType() <realm/Migration/deleteType.html>`.
```

This is useful if the previous version of a schema has a Realm object type,
but the new version of the schema does not.

.. literalinclude:: /examples/generated/flutter/migrations_test.snippet.migration-delete-type.dart
   :language: dart

.. _flutter-rename-property:

Rename a Property
~~~~~~~~~~~~~~~~~

Rename a schema property with :flutter-sdk:`Migration.renameProperty()
<realm/Migration/renameProperty.html>`.

.. literalinclude:: /examples/generated/flutter/migrations_test.snippet.migration-delete-type.dart
   :language: dart

.. _flutter-other-migration-tasks:

Other Migration Tasks
~~~~~~~~~~~~~~~~~~~~~

To perform other realm schema migrations, use the following properties of
the ``Migration`` object in your migration callback function:

> - :flutter-sdk:`Migration.newRealm <realm/Migration/newRealm.html>`: The Realm as it existed just before

Author: think so, good spot! will fix

  the migration with the previous schema version.
- :flutter-sdk:`Migration.oldRealm <realm/Migration/oldRealm.html>`:
  The Realm as it exists after the migration. By the end of the migration callback,
  all relevant data must be migrated from ``oldRealm`` into ``newRealm``.

> You must use the ``oldRealm``'s dynamic API to access its objects because you

Reviewer 3: I suggest we remove this paragraph, cause it's a bit confusing. The `Migration` class provides a special `MigrationRealm` instance that has limited set of APIs available while in the migration callback.

Reviewer 3: Lets remove if and just add this to the `oldRealm` description above
```
 /// The Realm as it existed just before the migration. Since the models have changed,
  /// this Realm can only be accessed via the dynamic API.
```

  The Realm as it exists after the migration. By the end of the migration callback,

>   all relevant data must be migrated from ``oldRealm`` into ``newRealm``.

Reviewer: ```suggestion
  you must migrate all relevant data  from ``oldRealm`` into ``newRealm``.
```

passive voice -> active

To find instances of an object in an old realm in the new realm,
use :flutter-sdk:`Migration.findInNewRealm() <realm/Migration/oldRealm.html>`.
To access the properties of objects from the old schema,
use the :flutter-sdk:`RealmObjects.dynamic <realm/RealmObject/dynamic.html>` API.

.. literalinclude:: /examples/generated/flutter/migrations_test.snippet.migrations-other.dart
   :language: dart

.. _flutter-update-schema-of-synced-realm:

Updating the Schema of a Synced Realm
-------------------------------------

Updating the schema of a synced realm is a separate process from updating

> the schema of a a local-only realm.

Reviewer 2: ```suggestion
the schema of a local-only realm.
```
